Do not generate -1* in SymPy expressions when canonicalising

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #128500
* __->__ #128411
* #128410

Partially addresses https://github.com/pytorch/pytorch/issues/128150

lezcano committed:
Do not generate -1* in SymPy expressions when canonicalising

Partially addresses https://github.com/pytorch/pytorch/issues/128150

[ghstack-poisoned]

lezcano committed:
Update on "Do not generate -1* in SymPy expressions when canonicalising"


Partially addresses https://github.com/pytorch/pytorch/issues/128150

[ghstack-poisoned]

lezcano committed:
Update on "Do not generate -1* in SymPy expressions when canonicalising"


Partially addresses https://github.com/pytorch/pytorch/issues/128150

[ghstack-poisoned]

lezcano committed:
Update on "Do not generate -1* in SymPy expressions when canonicalising"


Partially addresses https://github.com/pytorch/pytorch/issues/128150

[ghstack-poisoned]

lezcano commented on torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:
```diff
@@ -187,53 +187,41 @@ def _get_existing_inline_assertions(
-                maybe_symint_arg = maybe_symint_arg.args[1]
-                compare_op = operator.ge
-                compare_int = -1 * compare_int
+            lhs, rhs = compare_arg.args
```
@avikchaudhuri this needed quite a rewrite, but I think that now the code is much cleaner. PTAL.

tugsbayasgalan replied:
We are actually deprecating this pass very soon in favor of deferred runtime assertion pass. So as long as you make the test pass, i think it is ok to do whatever here.

lezcano commented on test/export/test_export.py:
```diff
@@ -2750,10 +2750,7 @@ def forward(self, x, y):
-            error_msg = r"Invalid value range for -1 between"
-        else:
-            error_msg = "is outside of inline constraint"
+        error_msg = r"Invalid value range for -1 between"
```
@avikchaudhuri not sure whether this change has any implications.

lezcano commented:
@ezyang now we have a nice-looking canonicalisation for inequalities :)

lezcano committed:
Update on "Do not generate -1* in SymPy expressions when canonicalising"


Partially addresses https://github.com/pytorch/pytorch/issues/128150

[ghstack-poisoned]

lezcano commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -365,23 +365,30 @@ def _canonicalize_bool_expr_impl(expr: SympyBoolean) -> SympyBoolean:
 
```
It is not though?
We still guarantee that. The only thing that has changed is that we prefer to leave things in the form `0 < smth` and, in general, `positive terms < other positive terms`.

ezyang replied:
Oh oops, you're right! Maybe we should add a little more text here.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -365,23 +365,30 @@ def _canonicalize_bool_expr_impl(expr: SympyBoolean) -> SympyBoolean:
+    def is_neg(t):
+        return t.is_negative or (isinstance(t, sympy.Mul) and t.args[0].is_negative)
+
+    lhs = 0
```
```
lhs >= rhs
~~~>
0 <= lhs - rhs
```

OK

```
lhs <= rhs
~~~~>
0 <= rhs - lhs
```

OK

ezyang replied:
TBH, it would be clearer what's going on if you move lhs up with the rhs assignment together, so it's clear that by line 378 (here) that we have equivalence

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -365,23 +365,30 @@ def _canonicalize_bool_expr_impl(expr: SympyBoolean) -> SympyBoolean:
-        rhs = -sympy.Add(*cts)
+                pos.append(term)
+        lhs = sympy.Add(*neg)
+        rhs = sympy.Add(*pos)
```
```
0 <= p0 + ... + pi + (-n0) + ... + (-nj)
~~~~>
n0 + ... nj <= p0 + ... + pi
```

OK

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -365,23 +365,30 @@ def _canonicalize_bool_expr_impl(expr: SympyBoolean) -> SympyBoolean:
+        rhs = sympy.Add(*pos)
+    elif is_neg(rhs):
+        # lhs == 0
+        lhs, rhs = -rhs, 0
```
```
0 <= (-n)
~~~~>
n <= 0
```

OK

lezcano committed:
Update on "Do not generate -1* in SymPy expressions when canonicalising"


Partially addresses https://github.com/pytorch/pytorch/issues/128150

[ghstack-poisoned]
